File: c/easy/find_smallest_letter_greater_than_target.c
#include <stddef.h>

char nextGreatestLetter(char* letters, int lettersSize, char target) {
    int left = 0, right = lettersSize;

    // binary search: upper_bound
    while (left < right) {
        int mid = left + (right - left) / 2;
        if (letters[mid] <= target)
            left = mid + 1;
        else
            right = mid;
    }

    return (left < lettersSize) ? letters[left] : letters[0];
}

File: rust/easy/find_smallest_letter_greater_than_target.rs
pub fn next_greatest_letter(letters: Vec<char>, target: char) -> char {
    let mut left = 0;
    let mut right = letters.len();

    while left < right {
        let mid = (left + right) / 2;
        if letters[mid] <= target {
            left = mid + 1;
        } else {
            right = mid;
        }
    }

    if left < letters.len() {
        letters[left]
    } else {
        letters[0]
    }
}
